src/app/(post-login)/submission/program/[shortName]/sample-registration/page.tsx

```diff
@@ -100,22 +100,82 @@ const Register = ({ shortName }: { shortName: string }) => {
 	const hasErrors = !!schemaOrValidationErrors.length;
 	const registrationId = get(clinicalRegistration, 'id', '') || '';
 
+	/**
+	 * new HTTP endpoints don't return the same objects as strictly typed gql endpoints
+	 * data shape needs to match so state can stay in sync on the frontend
+	 */
+	type RequiredProperty = { key: string; accessor: string | (() => void); defaultValue: any };
+	const requiredProperties: RequiredProperty[] = [
+		{
+			key: 'errors',
+			accessor: (resp) =>
+				resp.errors.map((error) => ({
+					type: error.type,
+					message: error.message,
+					row: error.index,
+					field: error.fieldName,
+					value: error.info.value,
+					sampleId: error.info.sampleSubmitterId,
+					donorId: error.info.donorSubmitterId,
+					specimenId: error.info.specimenSubmitterId,
+				})),
+			defaultValue: [],
+		},
+	];
+
+	const getValueFromAccessor = (data, accessor, defaultValue) => {
+		if (isFunction(accessor)) {
+			return accessor(data) || defaultValue;
+		} else if (isString(accessor)) {
+			return get(data, accessor, defaultValue);
+		}
+	};
+
+	// match a certain shape from gql
+	const satisfyGQL = ({
+		apiResp,
+		requiredProperties,
+	}: {
+		apiResp: any;
+		requiredProperties: RequiredProperty[];
+	}) =>
+		requiredProperties.reduce((acc, curr) => {
+			const value = getValueFromAccessor(apiResp, curr.accessor, curr.defaultValue);
+			return { ...acc, ...{ [curr.key]: value } };
+		}, {});
+
 	// handlers
 	const uploadURL = urlJoin(CLINICAL_API_ROOT, getProgramPath(UPLOAD_REGISTRATION, shortName));
 	const uploadFile = useMutation(
 		(formData) => {
 			return uploadFileRequest(uploadURL, formData, egoJwt);
 		},
 		{
-			onSuccess: (data, variables, context) => refetch(),
+			// success of query - not HTTP status
+			onSuccess: async (data, variables, context) => {
+				// succesful uploads will have state on the server
+				if ([200, 201].includes(data.status)) {
+					refetch();
+				} else {
+					// errors are not persisted server side - manually update cache (state store)
+					const registration = await data.json();
+					const registrationUpdate = satisfyGQL({ apiResp: registration, requiredProperties });
+
+					updateClinicalRegistrationQuery((previous) => {
```

Comment:
Will handlers like this be required for other requests?

Reply:
Yes. I believe so. With that in mind, I wrote these in a way that it should be easy to reuse

Reply:
Anything that is:
- not a GQL response
- doesn't have state server side (ie. can't invalidate and get new data with Apollo `refetch`)

Reply:
JS Docs are good, wondering if it's worth documenting in the readme, so there's a guide for future features where this is required